File: examples/payments/ledger/src/main/java/org/portolan/payments/ledger/domain/payment/vo/Money.java
package org.portolan.payments.ledger.domain.payment.vo;

import org.jmolecules.ddd.annotation.ValueObject;

/**
 * An amount in the minor unit of a currency: 1250 GBP is £12.50.
 *
 * The ledger never rounds. Everything it holds is an integer of minor units,
 * and a currency is compared before amounts are added at all.
 */
@ValueObject
public record Money(long amountMinor, String currency) {

    public Money {
        if (currency == null || currency.length() != 3) {
            throw new IllegalArgumentException("a currency is three letters of ISO 4217");
        }
    }

    public Money plus(Money other) {
        if (!currency.equals(other.currency)) {
            throw new IllegalArgumentException("cannot add " + other.currency + " to " + currency);
        }
        return new Money(amountMinor + other.amountMinor, currency);
    }

    public Money negated() {
        return new Money(-amountMinor, currency);
    }
}

File: examples/shop/pricing/internal/domain/quote/vo/money/money.go
// Package money is an amount in the minor unit of a currency.
package money

import "fmt"

// Money never rounds: everything is an integer of minor units, and two amounts
// are only added when they are in the same currency.
type Money struct {
	amountMinor int64
	currency    string
}

func New(amountMinor int64, currency string) (Money, error) {
	if len(currency) != 3 {
		return Money{}, fmt.Errorf("a currency is three letters of ISO 4217, got %q", currency)
	}

	return Money{amountMinor: amountMinor, currency: currency}, nil
}

func (m Money) AmountMinor() int64 { return m.amountMinor }

func (m Money) Currency() string { return m.currency }

func (m Money) Plus(other Money) (Money, error) {
	if m.currency != other.currency {
		return Money{}, fmt.Errorf("cannot add %s to %s", other.currency, m.currency)
	}

	return Money{amountMinor: m.amountMinor + other.amountMinor, currency: m.currency}, nil
}

func (m Money) Times(n int32) Money {
	return Money{amountMinor: m.amountMinor * int64(n), currency: m.currency}
}
